src/components/common/DiscountCode.vue

```diff
@@ -0,0 +1,123 @@
+<template>
+  <div class="row">
+    <div class="col-sm-offset-9 col-sm-3">
+      <ServerError :error="serverError">
+        <template slot-scope="{ graphQLError }">
+          {{ getErrorMessage(graphQLError) }}
+        </template>
+      </ServerError>
+    </div>
+    <div class="col-sm-12">
+      <form @submit.prevent="submit" class="text-right promotion-info">
+        <span class="text-uppercase promo-info-text">
+          {{ $t('promoInfo') }}
+          <img src="../../assets/img/information.png"
+               class="info-icon"
+               alt="information icon">
+        </span>
+        <input id="promo-code" type="text" name="code" v-model="code"
+               data-test="discount-code-input">
+        <LoadingButton :buttonState="buttonState"
+                       class="apply-button"
+                       data-test="apply-discount-code-button">
+          {{ $t('apply') }}
+        </LoadingButton>
+        <div v-if="cartLike.discountCodes.length > 0"
+             class="row"
+             style="margin-top: 15px">
+          <div class="col-sm-12">
+            <div class="text-right order-discount col-sm-offset-5">
+              <span class="col-sm-3"
+                    style="font-weight: bold">
+                Applied discounts:
+              </span>
+            </div>
+            <div v-for="discountInfo in cartLike.discountCodes"
+                :key='discountInfo.discountCode.id'
+                class="text-right order-discount col-sm-offset-7">
+              <div class="col-sm-3"
+                   data-test="discount-code-name">
+                {{ discountInfo.discountCode.code }}
+              </div>
+              <div class="col-sm-3">{{ discountInfo.discountCode.name }}</div>
+              <div class="col-sm-5">
+                ({{ discountInfo.discountCode.description }})
+              </div>
+              <div>
+                <div style="cursor: pointer"
+                    @click="removeDisountCode(discountInfo.discountCode.id)">
+                  <img src="../../assets/img/delete-1.png"
+                      class="cart-action-icon">
+                </div>
+              </div>
+            </div>
+          </div>
+        </div>
+      </form>
+    </div>
+  </div>
+</template>
+
+<script>
+import LoadingButton from './LoadingButton.vue';
+import ServerError from './ServerError.vue';
+
+export default {
+  components: { LoadingButton, ServerError },
+
+  props: ['buttonState', 'serverError', 'cartLike'],
+
+  data: () => ({
+    code: null,
+  }),
+
+  methods: {
+    submit() {
+      this.$emit('apply-code', this.code);
```

Comment:
Is there a reason why you are delegating the cart update to the parent? Why can't this component handle everything on its own? In the case of the LineItem component it was delegated because there were many forms (as many as line items) and only one cart, and it meant having to pass the cart id/version or to fetch it from each one of the components.

And actually I'm seriously considering not to delegate it anymore and fetching the cart id/version instead from inside LineItem component. It's not double-fetching because the information is cached, and this way we keep the whole domain wrapped in a component.